lib/ansible/modules/user.py

```diff
@@ -642,7 +642,10 @@ def create_user_useradd(self):
             if self.create_home:
                 parent = os.path.dirname(self.home)
                 if not os.path.isdir(parent):
-                    self.create_homedir(self.home)
+                    try:
+                        os.makedirs(parent)
```

Comment:
This fix should be in `create_homedir()`, or add the `-M` option to the command if `self.local`.

Reply:
Given your comment I made some further thoughts on the general approach for creating the home directory. If we leave alone the problem that caused this PR to be created (a missing parent directory that causes `luseradd` to fail currently) I see the following situation if `create_home` is set to `yes`:
1. If the user does not exist the home directory is created and filled with the configured skeleton contents by `useradd` / `luseradd`.
2. If the user exists and just the home directory is missing then `create_homedir()` performs this job to create the home directory and copy the skeleton contents. I guess this is the cause because `usermod` / `lusermod` do not offer this functionality.

I am not 100% sure if the home directory creation and skeleton copy by `useradd` / `luseradd` always deliver the same result as `create_homedir()`. I would not even be sure if `useradd` and `luseradd` always do the same.
Hence wouldn't it be better to
1. Never use the `-k` parameter on `useradd` / `luseradd`.
2. Never use the `-m` parameter for `useradd`.
3. Always set the `-M` parameter for `useradd` / `luseradd`.
4. Let `create_homedir()` always do the job of creating the home directory and copying the skeleton content if  `create_home` is set to `yes`.
This approach would also solve the issue the current patch tries to solve.

Reply:
Any comments?

Reply:
Is there anything I can do to move this forward or that is missing?

Reply:
The issue of home directory creation, particularly creating arbitrary parent directories, is quite tricky. I have been thinking about this off and on for several months and am still not clear on the “best” solution here.

In general, we really try to avoid creating arbitrary parent directories due to potential security issues. Any changes/additions to the creation of parent directories is alway something that requires a lot of careful consideration.

Compounding the situation are the increasingly apparent dirreneces between `luseradd` and `useradd`. There is a growing consensus on the Core Team that this module should be split into `local_user` and `user` modules to better handle these differences. The `local` toggle continues to create myriad issues and makes the code _much_ more complex.

Reply:
Sorry that's not terribly helpful. I'm working through the failure and trying to understand it a bit more in order to suggest an appropriate fix.

Reply:
> In general, we really try to avoid creating arbitrary parent directories due to potential security issues. Any changes/additions to the creation of parent directories is alway something that requires a lot of careful consideration.
> 
I would like to support here. Can you give me any hint which kind of potential security issues we could hit here provided their principles are already known to the general public and you can talk about them in the public? Is it related to race conditions between the parent directory creation and creating the home directory itself?

Reply:
> Can you give me any hint which kind of potential security issues we could hit here provided their principles are already known to the general public and you can talk about them in the public?

It's nothing secret. It's a fairly common attack scenario, and not in any way unique to Ansible.

If a malicious user knows the path their victim intends to create, they can create one or more of those parent directories with permissions that allow them to control sub directories, or sym linked into a directory they control.

Since the victim code is blindly creating all the directories in the path (essentially `mkdir -p`) without first checking each directory in that path for suspicious permissions or ACLs, it's possible a user's home directory could end up being accessible by an attacker.

A couple ways to thwart this attack include only creating parent directories from "trusted" roots, such as `/home` or checking, for each parent in the creation tree, if they exist and what their permissions are. The later case becomes terribly complex since some systems could have valid reasons why `/home` is 777, or there may be ACLs in place on `/home` providing some group access to all sub directories. There are just too many scenarios to come up with a reliable heuristic to detect "this directory has suspicious permissions" and then stoping execution. There would be too many false positives.

Reply:
But if I don't have a "trusted" root the same kind of issue can happen if I just have to create the home directory itself as the parent (or a component in the way up to '/') can be under the control of the attacker, correct? I guess if just a precreated home directory is under the attackers control this can be fixed by ensuring that its ownership and permissions are correct and that the ACL are correct / cleared. But as far as I can see other modules like ansible.builtin.file do not offer to set ACL's.

But I understand the issue that any existing path segments between a "trusted" prefix and a home directory are a potential issue and that the only safe way seems to be to create any of these segments on your own and not allow any prexisting "non trusted" segments. But this creates issues with idempotency as the "trusted" prefix which would need to be supplied by the user would need to grow after our initial execution. And as you said "detecting" a trusted path either results in false positives for insecure segments or leaves holes open.

And things become even more difficult when symlinks are added to this equation.

Another idea would be that the if the user specifies a "trusted" prefix every segment in between this and the target directory either doesn't exist or needs to have specific ownerships, permissions and ACL for the creation to succeed. But I guess this would be so awful complex to configure correctly from user perspective that it would not help either.
So maybe the best way would be to not support the creation of parent directories at all and require that the user precreates them via ansible.builtin.file before in a manner that is safe and appropriate for the specific use case.

OTOH it looks to me that ansible.builtin.file fights with similar issues and does not seem to have a silver bullet either.

Reply:
> So maybe the best way would be to not support the creation of parent directories at all and require that the user precreates them via ansible.builtin.file before in a manner that is safe and appropriate for the specific use case.

> OTOH it looks to me that ansible.builtin.file fights with similar issues and does not seem to have a silver bullet either.

Exactly. 😄 It's complicated and there's no easy solution.

Reply:
Which brings us back to the original issue: Does it make the security situation worse if we only do `os.makedirs` up to the level above the home directory and let `useradd` / `luseradd` create the home directory? Currently we create the whole path including the home directory itself either directly via  `os.makedirs` in case there is no skeleton directory or indirectly via `shutil.copytree` which uses `os.makedirs` internally.